### PR TITLE
Throw terminating error in New-TemporaryFile and make it not rely on the presence of the 'TEMP' environment variable

### LINKED ISSUE #4634: New-TemporaryFile should throw terminating error
As part of PR 4612, it was discussed [here](https://github.com/PowerShell/PowerShell/pull/4612#discussion_r134333394) that New-TemporaryFile should throw a terminating error (instead of a non-terminating one). Everyone was clear on the expected behaviour but the issue is that this change would be a breaking change in behaviour.

### PR DESCRIPTION
**Pull request by bergmeister**
## PR Summary

- Fixes #4634 by throwing a terminating error as agreed
- Makes the command not rely on the presence of the `TEMP` environment variable to get path to temp directory and use the .Net method `Path.GetTempPath()` instead
- Catch exception more specific as given by the documentation
- Improve existing test

## PR Checklist

Note: Please mark anything not applicable to this PR `NA`.

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - [x] Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] User facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - [ ] Issue filed - Issue link:
- [ ] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes) Calls to `New-TemporaryFile` that are in a pipeline will abort (by design)
- [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting) I have tested it locally but we would be rather testing .Net and test setup would be expensive because one needs to first create 65535  temp files
    - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.

### CODE REVIEW

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs [hunk 1]
```diff
@@ -10,7 +10,7 @@ namespace Microsoft.PowerShell.Commands
     /// <summary>
     /// The implementation of the "New-TemporaryFile" cmdlet
     /// </summary>
-    [Cmdlet(VerbsCommon.New, "TemporaryFile", SupportsShouldProcess = true, HelpUri = "https://go.microsoft.com/fwlink/?LinkId=526726")]
+    [Cmdlet(VerbsCommon.New, "TemporaryFile", SupportsShouldProcess = true, HelpUri = "https://docs.microsoft.com/powershell/module/microsoft.powershell.utility/new-temporaryfile")]
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs]: Please revert the change - we should use official short link.

File: test/powershell/engine/Help/assets/HelpURI/V3Cmdlets.csv [hunk 2]
```diff
@@ -63,7 +63,7 @@ Enter-PSHostProcess,https://go.microsoft.com/fwlink/?LinkId=403736
 Exit-PSHostProcess,https://go.microsoft.com/fwlink/?LinkId=403737
 Clear-RecycleBin,https://go.microsoft.com/fwlink/?LinkId=524082
 ConvertFrom-String,https://go.microsoft.com/fwlink/?LinkId=507579
-New-TemporaryFile,https://go.microsoft.com/fwlink/?LinkId=526726
+New-TemporaryFile,https://docs.microsoft.com/powershell/module/microsoft.powershell.utility/new-temporaryfile
```
Comment (iSazonov) [inline test/powershell/engine/Help/assets/HelpURI/V3Cmdlets.csv]: The same - please revert.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/NewTemporaryFile.Tests.ps1 [hunk 3]
```diff
@@ -14,17 +14,16 @@
...
-        Test-Path $tempFile | Should be $true
-        $tempFile | Should BeOfType System.IO.FileInfo
-
-        if(Test-Path $tempFile)
-        {
-            Remove-Item $tempFile -ErrorAction SilentlyContinue -Force
+            $tempFile | Should Exist
+            $tempFile | Should BeOfType System.IO.FileInfo
+        } finally {
+            [System.IO.File]::Delete($tempFile)
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/NewTemporaryFile.Tests.ps1]: Previous `Test-Path` and `Remove-Item` was better.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs (hunk 1 above)
Comment (bergmeister) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs]: Well, the official short link is broken.,,

File: test/powershell/Modules/Microsoft.PowerShell.Utility/NewTemporaryFile.Tests.ps1 (hunk 3 above)
Comment (bergmeister) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/NewTemporaryFile.Tests.ps1]: Why? The .net method is only one line because it internally performs the check if the file exists and does not throw in this case.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs (hunk 1 above)
Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs]: Feel free to open new Issue if the link is broken.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/NewTemporaryFile.Tests.ps1 (hunk 3 above)
Comment (iSazonov) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/NewTemporaryFile.Tests.ps1]: The method can throw https://msdn.microsoft.com/en-us/library/system.io.file.delete(v=vs.110).aspx?f=255&mspperror=-2147217396#Anchor_1

**Comment by bergmeister** [comment]
@iSazonov I addressed your comments and opened an issue for the help link [here](https://github.com/PowerShell/PowerShell/issues/6186).

**Comment by iSazonov** [comment]
@bergmeister Please update PR summary and title.

**Comment by bergmeister** [comment]
@iSazonov Ok. Done.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/NewTemporaryFile.Tests.ps1 [hunk 4]
```diff
@@ -14,17 +14,18 @@
...
-        $tempFile = New-TemporaryFile
-
-        Test-Path $tempFile | Should be $true
-        $tempFile | Should BeOfType System.IO.FileInfo
-
-        if(Test-Path $tempFile)
-        {
-            Remove-Item $tempFile -ErrorAction SilentlyContinue -Force
+        try {
+            $tempFile = New-TemporaryFile
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/NewTemporaryFile.Tests.ps1]: I think we should remove the `try` block - if the line throw we skip the test silently.
```Powershell
{ $script:tempFile = New-TemporaryFile } | Should -Not -Throw
$script:tempFile| Should Exist
...
```

Comment (bergmeister) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/NewTemporaryFile.Tests.ps1]: @iSazonov try-finally would still throw and not make the test skip silently due to the missing catch block but I also agree with you that in this special case, we do not need to use the try-finally pattern, so I will remove it

Comment (bergmeister) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/NewTemporaryFile.Tests.ps1]: Also, the `Should -Not -Throw` syntax is already deprecated in Pester 4.2: https://github.com/pester/Pester/releases/tag/4.2.0

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs (hunk 1 above)
Comment (SteveL-MSFT) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs]: Agree w/ @iSazonov, we should keep the temp link and have separate issue to fix that link.  cc @joeyaiello

File: test/powershell/Modules/Microsoft.PowerShell.Utility/NewTemporaryFile.Tests.ps1 [hunk 5]
```diff
@@ -14,16 +14,15 @@
...
 #>
 
-Describe "NewTemporaryFile" -Tags "CI" {
+Describe "New-TemporaryFile" -Tags "CI" {
 
     It "creates a new temporary file" {
         $tempFile = New-TemporaryFile
 
-        Test-Path $tempFile | Should be $true
+        $tempFile | Should Exist
```
Comment (SteveL-MSFT) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/NewTemporaryFile.Tests.ps1]: From a black box testing point of view, it seems that we should check that the $tempFile path starts with the path returned from GetTempPath()?

**Review by SteveL-MSFT — APPROVED** [review]
LGTM, one non-blocking comment

File: test/powershell/Modules/Microsoft.PowerShell.Utility/NewTemporaryFile.Tests.ps1 (hunk 5 above)
Comment (bergmeister) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/NewTemporaryFile.Tests.ps1]: You could... Would you be ok with this one:
```` powershell
$tempFile | Should BeLikeExactly "$([System.IO.Path]::GetTempPath())*"
````

File: test/powershell/Modules/Microsoft.PowerShell.Utility/NewTemporaryFile.Tests.ps1 (hunk 4 above)
Comment (iSazonov) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/NewTemporaryFile.Tests.ps1]: I see 4.2 introduce right behavior for `Should -Not -Throw `

**Review by iSazonov — APPROVED** [review]

**Comment by daxian-dbw** [comment]
@iSazonov The `[breaking-change]` tag should be added to the body of the commit message instead of the title of it. Sorry that I didn't point out that specifically.

When squash-merging a breaking change PR in future, please do like the following:
image

**Comment by iSazonov** [comment]
@daxian-dbw Thanks! Clear! Please re-check last commits which I merged.
